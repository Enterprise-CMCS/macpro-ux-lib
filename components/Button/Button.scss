@use "../../assets/theme/macbis-colors" as c;

.usa-button:disabled {
  cursor: default;
}

.cms-primary-background {
  background-color: c.$primary;
}

.cms-primary-text {
  color: c.$primary;
  box-shadow: 0 0 0 2px c.$primary;
}

.usa-button--secondary-outline {
  color: white;
  box-shadow: 0 0 0 2px white;
  background-color: transparent;

  &:hover {
    background-color: transparent;
    color: c.$gray-lightest;
  }

  &:active {
    background-color: #ffffff;
  }

  &:disabled,
  &:hover:disabled {
    background-color: transparent;
    color: c.$gray-light;
    box-shadow: 0 0 0 2px c.$gray-light;
  }
}

.usa-button--inverse {
  background-color: c.$gray-lighter;
  box-shadow: none;
  color: black;

  &:hover {
    background-color: c.$gray-lightest;
    color: black;
  }

  &:active {
    background-color: #ffffff;
  }

  &:disabled,
  &:hover:disabled {
    background-color: #5b616a;
    color: black;
  }
}
