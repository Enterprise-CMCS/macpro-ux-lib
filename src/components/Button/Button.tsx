import React from "react";
import { Icon, IconChoice } from "../Icon/Icon";

type IntrinsicElements = JSX.IntrinsicElements["button"];

interface Props extends IntrinsicElements {
  buttonText: string;
  buttonVariation?: ButtonVariation;
  disabled?: boolean;
  iconName?: IconChoice;
  shiftIconLeft?: boolean;
  ariaLabel?: string;
  largeButton?: boolean;
}

/**
 * Button Component
 * @param {string}      buttonText            Renders the text contained in the button.
 * @param {string}      [buttonVariation]     Renders the style of the button.
 * @param {boolean}     [disabled]            Determines whether or not a button is enabled.
 * @param {IconChoice}  [iconName]            Handles the icon to render in the button.
 * @param {boolean}     [shiftIconLeft]       Determines if a rendered icon should render on the left side.
 * @param {string}      [ariaLabel]           Overwrites the aria label of the button element that will be read on screen readers.
 * @param {boolean}     [largeButton]         Determines whether a large button is required.
 * @param {string}      [button]              Type of specified button.
 * @param {Event}       [onClick]             Handles its behavior when the button is clicked.
 * @param {string}      [target]              Specifies a name or a keyword that indicates where to display the response that is received after clicking the button.
 */

type ButtonVariation =
  | "primary"
  | "secondary"
  | "error"
  | "success"
  | "inverse"
  | "base"
  | "link"
  | "secondary-outline";

const ButtonVariationConversion: { [key: string]: string } = {
  primary: " cms-primary-background",
  secondary: "outline cms-primary-text",
  error: "secondary",
  success: " bg-green",
  inverse: "inverse",
  base: "base",
  link: "unstyled",
  "secondary-outline": "secondary-outline",
};

export const Button: React.FC<Props> = ({
  buttonVariation = "primary",
  buttonText,
  shiftIconLeft = false,
  iconName,
  ariaLabel,
  largeButton = false,
  disabled = false,
  ...rest
}) => {
  const buttonVariationType = ButtonVariationConversion[buttonVariation] ?? "";
  return (
    <button
      disabled={disabled}
      {...rest}
      aria-label={ariaLabel || `${buttonText} button`}
      className={`display-flex usa-button usa-button--${buttonVariationType} ${
        largeButton ? "usa-button--big" : ""
      }`}
    >
      {shiftIconLeft && iconName && (
        <Icon name={iconName} color="curentColor" />
      )}
      <span className={`padding-x-05${iconName ? " padding-top-05" : ""}`}>
        {buttonText}
      </span>
      {iconName && !shiftIconLeft && (
        <div>
          <Icon name={iconName} color="curentColor" />
        </div>
      )}
    </button>
  );
};
