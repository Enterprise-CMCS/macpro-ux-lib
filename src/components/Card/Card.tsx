import React, { PropsWithChildren } from "react";

interface Props {
  alignContent?: "left" | "right" | "center";
  altText?: string;
  bodyText?: string;
  exdent?: boolean;
  flagLayout?: boolean;
  flagRight?: boolean;
  headerText?: string;
  imageSource?: string;
  insetMedia?: boolean;
}

/**
 * Card Component
 * @param {string}    [alignContent]  String prop that aligns card content left, right, or center.
 * @param {string}    [altText]       Image alt text.
 * @param {string}    [bodyText]      Text for the card body.
 * @param {boolean}   [exdent]        Sets header at top of card, drops media below. This value will override flagLayout.
 * @param {boolean}   [flagLayout]    Controls the style of the card (normal or flag).
 * @param {boolean}   [flagRight]     Controls how the flag content is aligned.
 * @param {string}    [headerText]      Card header text.
 * @param {string}    [imageSource]   Source of image file to be used in card.
 * @param {boolean}   [insetMedia]    Set a border of whitespace around card content.
 */

export const Card: React.FC<PropsWithChildren<Props>> = ({
  alignContent,
  altText,
  bodyText,
  children,
  exdent,
  flagLayout,
  flagRight,
  headerText,
  imageSource,
  insetMedia,
  ...rest
}) => {
  const cardStyle = flagLayout ? "usa-card--flag" : "";
  const mediaRight = flagRight ? "usa-card--media-right" : "";
  const insetMediaMod = insetMedia ? "usa-card__media--inset" : "";

  return (
    <div className={`usa-card ${cardStyle} ${mediaRight}`} {...rest}>
      <div className="usa-card__container" style={{ textAlign: alignContent }}>
        <div className="usa-card__header">
          <h2 className="usa-card__heading">{headerText}</h2>
        </div>

        {imageSource && (
          <div
            className={`usa-card__media${
              exdent ? "--exdent" : ""
            } ${insetMediaMod}`}
          >
            <div className="usa-card__img">
              <img src={imageSource} alt={altText} />
            </div>
          </div>
        )}
        <div className="usa-card__body">
          <p className="usa-card__text">{bodyText}</p>
        </div>
        <div className="usa-card__footer">{children}</div>
      </div>
    </div>
  );
};
