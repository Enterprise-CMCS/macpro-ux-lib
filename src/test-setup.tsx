import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
