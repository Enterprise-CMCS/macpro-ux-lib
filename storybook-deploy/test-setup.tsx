import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };

/**
 * Clean up attributes from the rendered component by removing the generated ids.
 *
 * Before:
 *   for="input-type-text-123456"
 *   id="input-type-text-123456"
 *
 * After:
 *   for="input-type-text"
 *   id="input-type-text"
 *
 * This is useful during snapshot testing. Snapshot comparisons will fail without addressing these items.
 *
 * @param {HTMLElement} container The component to clean up.
 * @param {prefixes}      string[]  The reference substrings to leave alone.
 */
