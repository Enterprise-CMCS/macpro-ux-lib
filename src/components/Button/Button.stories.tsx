import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import { Button } from "./Button";

export default {
  title: "USWDS/Base/Button",
  component: Button,
  argTypes: {
    buttonVariation: {
      description: "Renders the style of the button",
    },
    buttonText: {
      description: "Renders the text contained in the button",
    },
    disabled: {
      control: "boolean",
      description: "Determines whether or not a button is disabled",
    },
    onClick: {
      description: "Handles button behavior when clicked",
      action: "clicked",
    },
    iconName: {
      description:
        "Determines which icon that needs to be rendered from: https://designsystem.digital.gov/components/icon/. Use the icon's listed name, for example: 'accessibility_new'",
    },
    shiftIconLeft: {
      description:
        "Determines if a rendered icon should render on the left side, the iconName prop needs to be filled",
    },
    largeButton: {
      description: "Determines whether a large button is required",
    },
    ariaLabel: {
      description:
        "Overwrites the aria label of the button element that will be read on screen readers",
    },
    type: {
      options: ["submit", "reset", "button"],
      control: { type: "select" },
      description: "Type of specified button",
    },
    target: {
      options: ["_blank", "_self", "_parent", "_top"],
      control: { type: "select" },
      description:
        "Specifies a name or a keyword that indicates where to display the response that is received after clicking the button",
    },
  },
  args: {
    buttonText: "Button",
    buttonVariation: "primary",
    disabled: false,
    onClick: undefined,
    iconName: undefined,
    shiftIconLeft: false,
    type: "button",
    target: "_self",
  },
} as ComponentMeta<typeof Button>;

const Template: ComponentStory<typeof Button> = ({ buttonText, ...rest }) => (
  <Button buttonText={buttonText} {...rest} />
);

export const PrimaryButton = Template.bind({});
PrimaryButton.args = {
  buttonVariation: "primary",
};

export const PrimaryButtonWithIcon = Template.bind({});
PrimaryButtonWithIcon.args = { buttonVariation: "primary", iconName: "add" };
export const SecondaryButton = Template.bind({});
SecondaryButton.args = { buttonVariation: "secondary" };
export const ErrorButton = Template.bind({});
ErrorButton.args = { buttonVariation: "error" };
export const SuccessButton = Template.bind({});
SuccessButton.args = { buttonVariation: "success" };
export const InverseButton = Template.bind({});
InverseButton.args = { buttonVariation: "inverse" };
export const BaseButton = Template.bind({});
BaseButton.args = { buttonVariation: "base" };
export const LinkButton = Template.bind({});
LinkButton.args = { buttonVariation: "link" };
export const BigButton = Template.bind({});
BigButton.args = { largeButton: true };
export const OutlineButton = Template.bind({});
OutlineButton.args = { buttonVariation: "outline" };
