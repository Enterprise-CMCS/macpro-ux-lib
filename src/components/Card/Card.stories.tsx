import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";

import { Card } from "./Card";
import { Button } from "../Button/Button";
import { TextInput } from "../TextInput/TextInput";

export default {
  title: "USWDS/Base/Card",
  component: Card,
  args: {
    headerText: "A Basic Card",
  },
  argTypes: {},
} as ComponentMeta<typeof Card>;

const Template: ComponentStory<typeof Card> = ({ children, ...rest }) => (
  <Card {...rest}>{children}</Card>
);

export const Default = Template.bind({});
export const BasicCard = Template.bind({});
export const FlagCard = Template.bind({});

Default.args = {
  bodyText: "It does not have much content.",
  altText: "Card Media Alt Text",
};

BasicCard.args = {
  bodyText: "This card has an image, a button, and a little text.",
  children: <Button>Button Text</Button>,
  imageSource:
    "https://designsystem.digital.gov/img/introducing-uswds-2-0/built-to-grow--alt.jpg",
};

FlagCard.args = {
  bodyText: "This card uses the flag layout, and uses the insetMedia prop.",
  children: (
    <TextInput
      label="This TextInput is a Child"
      fieldName="text-input-in-card"
    />
  ),
  flagLayout: true,
  imageSource:
    "https://designsystem.digital.gov/img/introducing-uswds-2-0/built-to-grow--alt.jpg",
  insetMedia: true,
};
