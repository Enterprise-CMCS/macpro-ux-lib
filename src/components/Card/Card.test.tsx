import React from "react";
import { screen, render } from "../../test-setup";
import { Button } from "../Button/Button";
import { Card } from "./Card";

describe("Card component", () => {
  it("Should render", () => {
    render(<Card headerText="Card Header Text" bodyText="Card Body Text" />);
    const card = screen.getByText("Card Header Text");
    expect(card).toBeInTheDocument();
  });

  it("Should render with children", () => {
    render(
      <Card headerText="Card Header Text" bodyText="Card Body Text">
        <Button>Visit</Button>
      </Card>
    );

    const button = screen.getByText("Visit");
    expect(button).toBeInTheDocument();
  });

  it("Should render with an image", () => {
    render(
      <Card
        headerText="Card Header Text"
        bodyText="Card Body Text"
        imageSource="https://designsystem.digital.gov/img/introducing-uswds-2-0/built-to-grow--alt.jpg"
        altText="Card Media Alt Text"
      />
    );

    const card = screen.getByText("Card Header Text");
    expect(card).toBeInTheDocument();
  });

  describe("component snapshots", () => {
    it("default", () => {
      const { container } = render(
        <Card headerText="Card Header Text" bodyText="Card Body Text" />
      );
      expect(container).toMatchSnapshot();
    });

    it("media and child component", () => {
      const { container } = render(
        <Card
          headerText="Card Header Text"
          bodyText="Card Body Text"
          imageSource="https://designsystem.digital.gov/img/introducing-uswds-2-0/built-to-grow--alt.jpg"
          altText="Card Media Alt Text"
        >
          <Button>Visit</Button>
        </Card>
      );
      expect(container).toMatchSnapshot();
    });

    it("media and child component in flag", () => {
      const { container } = render(
        <Card
          headerText="Card Header Text"
          bodyText="Card Body Text"
          imageSource="https://designsystem.digital.gov/img/introducing-uswds-2-0/built-to-grow--alt.jpg"
          altText="Card Media Alt Text"
          flagLayout
        >
          <Button>Visit</Button>
        </Card>
      );
      expect(container).toMatchSnapshot();
    });

    it("media and child component in flag layout right", () => {
      const { container } = render(
        <Card
          headerText="Card Header Text"
          bodyText="Card Body Text"
          imageSource="https://designsystem.digital.gov/img/introducing-uswds-2-0/built-to-grow--alt.jpg"
          altText="Card Media Alt Text"
          flagLayout
          flagRight
        >
          <Button>Visit</Button>
        </Card>
      );
      expect(container).toMatchSnapshot();
    });

    it("media with inset and header exdent", () => {
      const { container } = render(
        <Card
          headerText="Card Header Text"
          bodyText="Card Body Text"
          imageSource="https://designsystem.digital.gov/img/introducing-uswds-2-0/built-to-grow--alt.jpg"
          altText="Card Media Alt Text"
          insetMedia
          exdent
        >
          <Button>Visit</Button>
        </Card>
      );
      expect(container).toMatchSnapshot();
    });
  });
});
