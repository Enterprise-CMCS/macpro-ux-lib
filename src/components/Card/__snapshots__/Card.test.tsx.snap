// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component component snapshots default 1`] = `
<div>
  <div
    class="usa-card  "
  >
    <div
      class="usa-card__container"
    >
      <div
        class="usa-card__header"
      >
        <h2
          class="usa-card__heading"
        >
          Card Header Text
        </h2>
      </div>
      <div
        class="usa-card__body"
      >
        <p
          class="usa-card__text"
        >
          Card Body Text
        </p>
      </div>
      <div
        class="usa-card__footer"
      />
    </div>
  </div>
</div>
`;

exports[`Card component component snapshots media and child component 1`] = `
<div>
  <div
    class="usa-card  "
  >
    <div
      class="usa-card__container"
    >
      <div
        class="usa-card__header"
      >
        <h2
          class="usa-card__heading"
        >
          Card Header Text
        </h2>
      </div>
      <div
        class="usa-card__media "
      >
        <div
          class="usa-card__img"
        >
          <img
            alt="Card Media Alt Text"
            src="https://designsystem.digital.gov/img/introducing-uswds-2-0/built-to-grow--alt.jpg"
          />
        </div>
      </div>
      <div
        class="usa-card__body"
      >
        <p
          class="usa-card__text"
        >
          Card Body Text
        </p>
      </div>
      <div
        class="usa-card__footer"
      >
        <button
          aria-label="Visit button"
          class="display-flex usa-button usa-button-- "
        >
          <span
            class="padding-x-05"
          >
            Visit
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Card component component snapshots media and child component in flag 1`] = `
<div>
  <div
    class="usa-card usa-card--flag "
  >
    <div
      class="usa-card__container"
    >
      <div
        class="usa-card__header"
      >
        <h2
          class="usa-card__heading"
        >
          Card Header Text
        </h2>
      </div>
      <div
        class="usa-card__media "
      >
        <div
          class="usa-card__img"
        >
          <img
            alt="Card Media Alt Text"
            src="https://designsystem.digital.gov/img/introducing-uswds-2-0/built-to-grow--alt.jpg"
          />
        </div>
      </div>
      <div
        class="usa-card__body"
      >
        <p
          class="usa-card__text"
        >
          Card Body Text
        </p>
      </div>
      <div
        class="usa-card__footer"
      >
        <button
          aria-label="Visit button"
          class="display-flex usa-button usa-button-- "
        >
          <span
            class="padding-x-05"
          >
            Visit
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Card component component snapshots media and child component in flag layout right 1`] = `
<div>
  <div
    class="usa-card usa-card--flag usa-card--media-right"
  >
    <div
      class="usa-card__container"
    >
      <div
        class="usa-card__header"
      >
        <h2
          class="usa-card__heading"
        >
          Card Header Text
        </h2>
      </div>
      <div
        class="usa-card__media "
      >
        <div
          class="usa-card__img"
        >
          <img
            alt="Card Media Alt Text"
            src="https://designsystem.digital.gov/img/introducing-uswds-2-0/built-to-grow--alt.jpg"
          />
        </div>
      </div>
      <div
        class="usa-card__body"
      >
        <p
          class="usa-card__text"
        >
          Card Body Text
        </p>
      </div>
      <div
        class="usa-card__footer"
      >
        <button
          aria-label="Visit button"
          class="display-flex usa-button usa-button-- "
        >
          <span
            class="padding-x-05"
          >
            Visit
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Card component component snapshots media with inset and header first 1`] = `
<div>
  <div
    class="usa-card  "
  >
    <div
      class="usa-card__container"
    >
      <div
        class="usa-card__header"
      >
        <h2
          class="usa-card__heading"
        >
          Card Header Text
        </h2>
      </div>
      <div
        class="usa-card__media--exdent usa-card__media--inset"
      >
        <div
          class="usa-card__img"
        >
          <img
            alt="Card Media Alt Text"
            src="https://designsystem.digital.gov/img/introducing-uswds-2-0/built-to-grow--alt.jpg"
          />
        </div>
      </div>
      <div
        class="usa-card__body"
      >
        <p
          class="usa-card__text"
        >
          Card Body Text
        </p>
      </div>
      <div
        class="usa-card__footer"
      >
        <button
          aria-label="Visit button"
          class="display-flex usa-button usa-button-- "
        >
          <span
            class="padding-x-05"
          >
            Visit
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
